Option to auto pulp or butcher corpses

OzoneH3 opened the pull request:
![](https://user-images.githubusercontent.com/3514404/31864570-c04b6fe8-b75f-11e7-9ce2-4f75cbc400c2.png)

Coolthulhu commented on src/game.cpp:
```diff
@@ -11724,6 +11724,49 @@ void game::place_player( const tripoint &dest_loc )
+                }
+            }
+        } else if ( pulp_butcher == "pulp" || pulp_butcher == "pulp_adjacent" ) {
+            static auto pulp = [&]( const tripoint &pos ) {
```
As far as I recall, lambdas should be declared constant where possible, otherwise they may be treated differently by the compiler or fail to compile on older compilers.

Coolthulhu commented on src/game.cpp:
```diff
@@ -11724,6 +11724,49 @@ void game::place_player( const tripoint &dest_loc )
+            if( pulp_butcher == "pulp_adjacent" ) {
+                static const direction adjacentDir[8] = { NORTH, NORTHEAST, EAST, SOUTHEAST, SOUTH, SOUTHWEST, WEST, NORTHWEST };
+                for( auto &elem : adjacentDir ) {
+                    pulp( tripoint( direction_XY( elem ), 0 ) + u.pos() );
```
Would be cleaner as `pulp( u.pos() + direction_XY( elem ) )`.

Coolthulhu commented:
* It automatically starts pulping acidic corpses when the player is not immune. This is dangerous and must be fixed.
* There is no way to quickly turn it off and on. You want it off whenever you can quickly lose the sight of enemies and don't want to suddenly start pulping after cutting a corner. Could be tied to safe mode.
* The adjacent autopulp needs some more info - currently it's hard to tell what was pulped, how many corpses were pulped and so on. Would be useful for it to say "Finished pulping the Zombie corpse (NW)".
* Styling

Coolthulhu commented on src/game.cpp:
```diff
@@ -11724,6 +11724,49 @@ void game::place_player( const tripoint &dest_loc )
+            static auto pulp = [&]( const tripoint &pos ) {
+                for( const auto &maybe_corpse : m.i_at( pos ) ) {
+                    if ( maybe_corpse.is_corpse() && maybe_corpse.damage() < maybe_corpse.max_damage() &&
+                        maybe_corpse.get_mtype()->has_flag( MF_REVIVES ) ) {
```
`item::can_revive()`

kevingranade commented:
I'm sceptical that an option to auto-pulp when safe mode is on is
necessary. If you've turned safe mode on, that's a pretty clear signal that
sporadically triggering long-duration actions isnt desired.

OzoneH3 commented:
@kevingranade The pulp safe-mode is not tied to the '!' safemode directly. It works the same way the autopickup safemode. If there are enemies in safemode proximity distance it just won't pulp anything.

> There is no way to quickly turn it off and on. You want it off whenever you can quickly lose the sight of enemies and don't want to suddenly start pulping after cutting a corner. Could be tied to safe mode.

Is this really neccesary? You have the auto pulp safemode option explained above and the normal zombie interrupting activity prompt when you spot an enemie while perfoming pulping or butchering.
Would also require another option entry and a keybinding thats better used for other things.
   
> The adjacent autopulp needs some more info - currently it's hard to tell what was pulped, how many corpses were pulped and so on. Would be useful for it to say "Finished pulping the Zombie corpse (NW)".

I'm not sure this is possible since the assigned activity is actually printing the pulp messages.

cainiaowu commented:
Unrelated but could we have auto harvest mode like this as next step? It could share lots code like safemode etc. just different activity.

Coolthulhu commented:
> Is this really neccesary?

Imagine the following situations:
* Fighting shady zombies, trying to relocate, no zombies in sight but multiple ones close by.
* Got boomer bile'd, trying to relocate, but forced into pulping every step.
* Just fought off a smoker and stepping from a smoke tile off to a smoke-free one,

It certainly deserves a quick toggle option more than autopickup does. I can't think of a situation where I'd like one off and one on, so maybe tying those two together would work.

kevingranade commented:
> @kevingranade The pulp safe-mode is not tied to the '!' safemode directly. It works the same way the autopickup safemode. If there are enemies in safemode proximity distance it just won't pulp anything.

This doesn't address my question, do we really need to support auto-pulping when there are monsters nearby? Is that a remotely sensible thing to do?

OzoneH3 commented:
> This doesn't address my question, do we really need to support auto-pulping when there are monsters nearby? Is that a remotely sensible thing to do?

Yes, somehow ignored the long action interrupt part and yes with that in mind it's just pointless.

Fixed and changed:
* Removed the safe mode option. (Always off as long an enemy is in sight)
* Added an on/off toggle keybinding/action.
* No longer pulps acidic zombies.

Things I don't know how to adress properly because it's in a long action:
> The adjacent autopulp needs some more info - currently it's hard to tell what was pulped, how many corpses were pulped and so on. Would be useful for it to say "Finished pulping the Zombie corpse (NW)".

Most of the time it will be only one corpse and pretty obvious where it was considering blood and bile flying around. Other than that I don't think a direction would be all that useful if there is a ton of corpses.
